crates/red_knot_python_semantic/src/types.rs

```diff
@@ -463,6 +463,55 @@ impl<'db> Type<'db> {
         self == other
     }
 
+    /// Return true if there is just a single inhabitant for this type.
+    ///
+    /// Note: This function aims to have no false positives, but might return `false`
+    /// for more complicated types that are actually singletons.
+    pub(crate) fn is_singleton(self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::Any
+            | Type::Never
+            | Type::Unknown
+            | Type::Unbound
+            | Type::Module(..)
+            | Type::Instance(..)
+            | Type::IntLiteral(..)
+            | Type::StringLiteral(..)
+            | Type::BytesLiteral(..)
+            | Type::LiteralString => {
+                // Note: The literal types included in this pattern are not true singletons.
+                // There can be multiple Python objects (at different memory locations) that
+                // are both of type Literal[345], for example.
+                false
+            }
+            Type::None | Type::BooleanLiteral(_) | Type::Function(..) | Type::Class(..) => true,
+            Type::Tuple(tuple) => {
+                !tuple.elements(db).is_empty()
+                    && tuple.elements(db).iter().all(|ty| ty.is_singleton(db))
+            }
+            Type::Union(..) => {
+                // There are some rare edge cases where a union type might be a singleton type.
+                // For example, a union with just one element (which itself is a singleton). Or
+                // a union with an empty type (e.g. Never | None). Here, we assume that such
+                // types would have been simplified to a different representation earlier and
+                // simply return false.
+                false
+            }
+            Type::Intersection(..) => {
+                // Intersection types are hard to analyze. The following types are technically
+                // all singleton types, but it is not straightforward to compute this. Again,
+                // we simply return false.
+                //
+                //   bool & ~Literal[False]`
+                //   None & (None | int)
+                //   (A | B) & (B | C)          with A, B, C disjunct and B a singleton
+                //
+                false
+            }
+            Type::Todo => todo!(),
```

Comment:
What rust-analyzer auto-inserted looks so logical, I just kept it :smile:. But seriously, I don't know how to treat `Type::Todo` here. Returning `false` might be another sensible option.